File: backend/app/schemas/timetable.py
"""
Pydantic schemas for Timetable.
"""

from datetime import date, time
from typing import Any, Optional

from pydantic import BaseModel


class PeriodTime(BaseModel):
    period: int
    start: str   # "HH:MM"
    end: str     # "HH:MM"


class TimetableConfigCreate(BaseModel):
    periods_per_day: int
    enable_weekends: bool = False
    period_times: Optional[list[Any]] = None


class TimetableConfigResponse(BaseModel):
    id: int
    periods_per_day: int
    enable_weekends: bool
    period_times: Optional[list[Any]] = None
    created_by_admin_id: Optional[int] = None

    model_config = {"from_attributes": True}


class TimetableSlotCreate(BaseModel):
    grade: int
    slot_date: date          # specific calendar date
    period_number: int
    subject: str
    teacher_id: Optional[int] = None
    start_time: Optional[time] = None
    end_time: Optional[time] = None
    is_holiday: bool = False
    comment: Optional[str] = None


class TimetableSlotUpdate(BaseModel):
    subject: Optional[str] = None
    teacher_id: Optional[int] = None
    start_time: Optional[time] = None
    end_time: Optional[time] = None
    is_holiday: Optional[bool] = None
    comment: Optional[str] = None


class TimetableSlotResponse(BaseModel):
    id: int
    grade: int
    slot_date: date
    period_number: int
    subject: str
    teacher_id: Optional[int] = None
    start_time: Optional[time] = None
    end_time: Optional[time] = None
    is_holiday: bool
    comment: Optional[str] = None

    model_config = {"from_attributes": True}


class TimetableSlotWithTeacherResponse(BaseModel):
    """Enriched slot response that includes the teacher's username and period
    times resolved from the admin timetable config."""
    id: int
    grade: int
    slot_date: str           # "YYYY-MM-DD" string
    period_number: int
    subject: str
    teacher_id: Optional[int] = None
    teacher_username: Optional[str] = None
    start_time: Optional[str] = None   # "HH:MM" string
    end_time: Optional[str] = None
    is_holiday: bool
    comment: Optional[str] = None

File: frontend/lib/core/models/timetable.dart
class TimetableSlotModel {
  final int id;
  final int grade;
  final String slotDate;   // "YYYY-MM-DD" — specific calendar date
  final int periodNumber;
  final String? subject;
  final int? teacherId;
  final String? teacherUsername;
  final String? startTime;
  final String? endTime;
  final bool isHoliday;
  final String? comment;

  const TimetableSlotModel({
    required this.id,
    required this.grade,
    required this.slotDate,
    required this.periodNumber,
    this.subject,
    this.teacherId,
    this.teacherUsername,
    this.startTime,
    this.endTime,
    required this.isHoliday,
    this.comment,
  });

  factory TimetableSlotModel.fromJson(Map<String, dynamic> json) =>
      TimetableSlotModel(
        id: json['id'] as int,
        grade: json['grade'] as int,
        slotDate: json['slot_date'] as String,
        periodNumber: json['period_number'] as int,
        subject: json['subject'] as String?,
        teacherId: json['teacher_id'] as int?,
        teacherUsername: json['teacher_username'] as String?,
        startTime: json['start_time'] as String?,
        endTime: json['end_time'] as String?,
        isHoliday: json['is_holiday'] as bool,
        comment: json['comment'] as String?,
      );
}

class TimetableConfigModel {
  final int id;
  final int periodsPerDay;
  final bool enableWeekends;
  final List<dynamic>? periodTimes;
  final int? createdByAdminId;

  const TimetableConfigModel({
    required this.id,
    required this.periodsPerDay,
    required this.enableWeekends,
    this.periodTimes,
    this.createdByAdminId,
  });

  factory TimetableConfigModel.fromJson(Map<String, dynamic> json) =>
      TimetableConfigModel(
        id: json['id'] as int,
        periodsPerDay: json['periods_per_day'] as int,
        enableWeekends: json['enable_weekends'] as bool,
        periodTimes: json['period_times'] as List<dynamic>?,
        createdByAdminId: json['created_by_admin_id'] as int?,
      );
}
